### ISSUE TITLE
[EPIC] Support cron strings as the primary scheduling method in the backend

### ISSUE THREAD

**Issue by jrhizor** [airbytehq/airbyte#2170]
This is a partial prerequisite for https://github.com/airbytehq/airbyte/issues/1965 on the frontend, but resolves it for the backend when users are calling the API.

See [this](https://github.com/airbytehq/airbyte/issues/2170#issuecomment-1051065420) for summary of approach.

Issues to tackle, in order:
- [x] #11418 
- [x] #11419 
- [x] #11431 
- [x] #15593
- [ ] #11432 

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368090097272) by [Unito](https://www.unito.io)

**Comment by michel-tricot** [comment]
How big is this project? We have one design partner that will need it.

**Comment by cgardens** [comment]
It's not trivial. Temporal isn't handling cron schedules right now, so that needs to happen first. This essentially means we need to handle Temporal metadata alot better. It's worth doing but we haven't gotten to it yet.

**Comment by cgardens** [comment]
I suspect it is a 3 days.

**Comment by erichartono** [comment]
This would be a handy feature to have. I need to run a daily sync at certain AM hour. Thanks.

**Comment by harshithmullapudi** [comment]
+1 from me. This would be a handy feature to have. I need to run a daily sync at certain AM hour. Is there any timeline for this?

**Comment by nathan5280** [comment]
I've used this Python package on other projects to help with building cron schedules.   https://pypi.org/project/croniter/

**Comment by danieldiamond** [comment]
+1 this one would be ace

**Comment by qiraahmad** [comment]
Would love it if this is implemented asap, if your transformations are interdependent this is a huge help!

**Comment by boggdan95** [comment]
Any news about the implementation of this?

**Comment by peder1001** [comment]
+1

**Comment by ShenW123** [comment]
+1

**Comment by callumkift** [comment]
+1

**Comment by thomasclavet** [comment]
+1

**Comment by bashyroger** [comment]
+1

**Comment by MaxKrog** [comment]
+1

**Comment by ameyabapat-bsft** [comment]
+1

**Comment by pranavhegde4** [comment]
+1, would help a lot in our ingestion pipeline if we could have more control over the scheduling

**Comment by bashyroger** [comment]
+1, the fact that this is (still) not supported now really makes scheduling cumbersome. This kind of functionality has always been there from the start in SAAS based El tools like **Fivetran** and **Stitch**. We now have to do 'extra work' to make this possible.

**Comment by vmgustavo** [comment]
Is there anything we could do as a temporary solution?

**Comment by marcosmarxm** [comment]
> Is there anything we could do as a temporary solution?

Use Manual Sync Frequency and trigger the sync using the API.

**Comment by ameyabapat-bsft** [comment]
+1. It is critical need for all scheduling related business use cases. Eagerly waiting for it.

**Comment by ameyabapat-bsft** [comment]
+1. It is a critical need for schedulers business use cases. Eagerly waiting for it.

**Comment by BEIIKS** [comment]
> > Is there anything we could do as a temporary solution?
> 
> Use Manual Sync Frequency and trigger the sync using the API.

you can make an API request to airbyte for a manual sync?

**Comment by marcosmarxm** [comment]
> > > Is there anything we could do as a temporary solution?
> > 
> > 
> > Use Manual Sync Frequency and trigger the sync using the API.
> 
> you can make an API request to airbyte for a manual sync?

Yes, it's possible.

**Comment by timroes** [comment]
Couple of open questions around this:

* Do we want to have a minimal interval allowed via a cron expression? By default using a cron expression you'd be able to run something up to every minute. I am wondering if we might want to apply some higher limit on that for safety reasons, or if we're fine having jobs run every minute (and whether we want to have different minimal intervals in OSS vs Cloud)?
* By default we're skipping the next sync when the previous one is still running at the time it should happen (from my understanding). While this might be okay with "Every 5 minutes", I wonder if there might be a different expectations when using a cron expression, that it _actually_ runs at the set times. E.g. when you said a cron expression for `0 3,5 * * *`, and the job runs 2h5mins at 3 o clock the 5 o clock one will simply be skipped (also no job logged for it), and the next one runs at 3 the next day. I wonder if there are higher expectations to explicit execution times when using cron expressions than "every x" intervals.

**Comment by timroes** [comment]
Talking about this, we decided we don't want to limit the minimum interval, but allow this on a per minute basis as specified via cron. It would be nice if we send out a notification if a cron task get skipped, but not a requirement for the first implementation.

**Comment by GKTheOne** [comment]
@timroes could the default 'when the sync fails' webhook notification work as a MVP for skipping (detection), instead of silently ignoring?

**Comment by timroes** [comment]
> could the default 'when the sync fails' webhook notification work as a MVP for skipping (detection), instead of silently ignoring?

Theoretically yes we can enable this, but this will require additional complexity to figure out, which we currently didn't see as a requirement to block this on (though obviously nice to have also for the first itteration).

**Comment by timroes** [comment]
Summary from the offline architectural discussion around how to implement this.

We'll be adding a `scheduleType: enum{ Manual, BasicSchedule, Cron }` flag into a connection, which will replace the existing `manual: boolean` flag. Depending on that flag the `schedule` field will contain a different format:

* `Manual`: `null`
* `BasicSchedule`: `{ timeUnits: TimeUnit, units: number }` (the existing format)
* `Cron`: `{ cronExpression: string, cronTimeZone: string }`

We feel this gives the fastes implementation and easiest migration for now, while still giving us flexibility of switching over to having a cron expression as the single source of truth for scheduling in the server, if we want to make that change. It will also allow us to later on add more complex scheduling, like "only run in specific time frames, etc."

The API would return (and accept) this in the format (slightly simplified OpenAPI format):

```
required: [scheduleType]
properties:
  scheduleType:
    type: string
    enum:
      - manual
      - basicSchedule
      - cron
  scheduleCron:
    required: [cronExpression, cronTimeZone]
    properties:
      cronExpression: string
      crontTimeZone: string
  scheduleBasic:
    required: [timeUnits, units]
  properties:
    timeUnits: TimeUnit
    units: number
```

For calculating cron expressions we should use [quartz](https://www.quartz-scheduler.org/api/1.8.6/org/quartz/CronTrigger.html) since it's lightweight enough and has support for time zones.

**Comment by misteryeo** [comment]
**Issue was linked to Harvestr Discovery:**  [Cron expression syncs](https://app.harvestr.io/components/0/list/1MfiXo0kS)

**Comment by TSkrebe** [comment]
Since there has been some activity recently regarding this topic, is there any timeline when we could this feature come to fruition?

**Comment by davinchia** [comment]
@supertopher will start working on this in the next few weeks. If all goes well, this should be out within a month or so (no promises).

**Comment by cgardens** [comment]
@supertopher we aren't working on this now, right? can i unassign you?

**Comment by ZiggerZZ** [comment]
So the [documentation](https://airbyte-public-api-docs.s3.us-east-2.amazonaws.com/rapidoc-api-docs.html#post-/v1/connections/create) has already included `scheduleType` and `scheduleData` fields in the schema, but when I include them in the request I get 
```
message: 'Invalid json input. Unrecognized field "scheduleData" (class io.airbyte.api.model.generated.ConnectionCreate), not marked as ignorable
(12 known properties: "operationIds", "namespaceFormat", "sourceId", "schedule", "destinationId", "name", "namespaceDefinition", "prefix", "sourceCatalogId", "syncCatalog", "status", "resourceRequirements"])\n' +
        
' at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 1, column: 655] (through reference chain: io.airbyte.api.model.generated.ConnectionCreate["scheduleData"])
Unrecognized field "scheduleData" (class io.airbyte.api.model.generated.ConnectionCreate), not marked as ignorable',
```

**Comment by davinchia** [comment]
@mfsiega-airbyte ^

**Comment by mfsiega-airbyte** [comment]
Re https://github.com/airbytehq/airbyte/issues/2170#issuecomment-1228504416 there isn't a ton to go off of so if there's any more debug info to reproduce I can take a closer look. In the meantime I am in the process of adding an explicit acceptance test so that will give some reassurance/clarity.

**Comment by andyjih** [comment]
https://github.com/airbytehq/airbyte/pull/16375

**Comment by nataliekwong** [comment]
Closing this issue as we currently support Cron as a scheduling method in both the UI and the [API](https://reference.airbyte.com/reference/createconnection#scheduling-customization). You can read more about how to use it in our [docs](https://docs.airbyte.com/using-airbyte/core-concepts/sync-schedules#cron-scheduling).  If you have any issues related to Cron scheduling beyond version v0.50.46, feel free to create a new GitHub issue.
